rfcs/20210201-resampling-with-no-pixel-shift/README.md

```diff
@@ -0,0 +1,112 @@
+Unfortunately, despite this option, some users still used the old inaccurate version,
+which remained the default version up to TF version 2.x.
+
+The results of the linear algorithm for oneDNN and the different versions of this algorithm 
+in TF are presented below. The following results are based on downsampling the image from
+3x3 to 2x2, where the original image was filled with the values: [[0,1,2], [3,4,5], [6,7,8]]
+
+![](oneDNN_result.png)
+
+![](TF_results.png)
+
+## Proposal
+
+### Option 1 Adding the resampling flags in the descriptor constructor:
+
+``` cpp
+/// Flags for resampling primitive.
+typedef enum {
+    dnnl_resampling_flags_none = 0x0U,
+    dnnl_half_pixel_centers = 0x1U,
+} dnnl_resampling_flags_t;
+```
+
+``` cpp
+/// Flags for resampling primitive.
+enum class resampling_flags : unsigned {
+    none = dnnl_resampling_flags_none,
+    half_pixel_centers = dnnl_half_pixel_centers
+};
+```
+
+``` cpp
+...
+desc(prop_kind aprop_kind, algorithm aalgorithm, resampling_flags flags,
+        const memory::desc &src_desc, const memory::desc &dst_desc)
+...
+desc(prop_kind aprop_kind, algorithm aalgorithm, resampling_flags flags, 
+        const std::vector<float> &factors, const memory::desc &src_desc)
+...
+desc(prop_kind aprop_kind, algorithm aalgorithm, resampling_flags flags,
+        const std::vector<float> &factors, const memory::desc &src_desc, const memory::desc &dst_desc)
+...
+```
+
+To maintain backward compatibility, instead of adding the resampling flags
+to existing constructors it would be needed to increase the number
+of constructors available for the resampling primitive descriptor. For
+existing constructors, a flag would have half_pixel_centers as the default value.
+
+Pros:
+- Consistency attribute with TensorFlow 1.x.
+- The possibility of using this information for each algorithm we support.
+- Possibility to add additional flags for the resampling primitive in the future.
+
+Cons:
+- API change for resampling primitive.
+- More complicated in implementation than option 2.
+- In case of using C API to avoid ABI break we need to add another version of 
+init function(similar to pooling primitive - version v1 and v2).
+
+### Option 2 Adding a new algorithm for resampling, linear_no_pixel_shift:
+
+``` cpp
+typedef enum {
+    ...
+    dnnl_resampling_linear_no_pixel_shift = number;
+    ...
+} dnnl_alg_kind_t;
+```
+
+Pros:
+- Change relied on increasing the number of available algorithms. This approach does
+not require so many changes in API as it is done in Option 1.
+- Separation of an implementation that is considered inaccurate.
+
+Cons:
+- In case of a request to add support for another algorithm with that behavior,
+we would have to add another algorithm to the list(example: algorithm nearest neighbour without
+half_pixel_centers behavior).
+
+### Recommendation
+I would recommend Option 2. The flag approach seems to be better since it makes
```

Comment:
Personally, I would vote for option number 1. In my opinion it is more extendable taking into consideration future development of resampling. 

What if in the future someone will request another feature ? For example align corners from https://www.tensorflow.org/api_docs/python/tf/compat/v1/image/resize_bilinear.

In option 1 - it will be another enum dnnl_align_corners=0x2. What is more it will be easy to combine features. To enable no_pixel_shift with align corners client will set bit 0 and 1. Additionally features will be available at once for all resampling algorithms linear and nearest_neighbour.
In option 2 adding features and combining it with algorithms requires adding a lot of enumerations - algorithm_name_feature_1_feature_2....feature_n.

Reply:
I agree option 1 is more flexible, but with that flexibility comes more complexity (see the batchnorm precedent).
I am personally in favor of option 2 since it will be simpler to implement/test, and because we are not planning to provide more extensions to resampling (TF2 removed support for this resampling flavor, and other frameworks do not use this 
 definition as it is "broken").

Here, because this oneDNN feature extension is about supporting legacy use cases, I would favor simplicity over flexibility. If it turns out we end up with more flavors over time, we can still consolidate them at that point.

Reply:
While initially I proposed option 1, i would agree with Mourad and vote for option 2. If it affects us in the future and we will see more properties required to be supported we can have resampling_v2.

Right now the second option is straight-forward {API, implementation} -wise and we have a plan for the case in which we need to extend API.

Sorry for the burden.

Reply:
Since there is no recent activity in this thread, I will assume that we are ok with option 2.